server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1255,15 +1256,59 @@ private void finalizeSnapshotEntry(SnapshotsInProgress.Entry entry, Metadata met
             }
             metadataListener.whenComplete(meta -> {
                         final Metadata metaForSnapshot = metadataForSnapshot(entry, meta);
-                        final SnapshotInfo snapshotInfo = new SnapshotInfo(snapshot.getSnapshotId(),
+
+                        final Map<String, SnapshotInfo.IndexSnapshotDetails> indexSnapshotDetails = new HashMap<>(finalIndices.size());
+                        for (ObjectObjectCursor<ShardId, ShardSnapshotStatus> shardEntry : entry.shards()) {
+                            indexSnapshotDetails.compute(shardEntry.key.getIndexName(), (indexName, current) -> {
+                                if (current != null && current.getShardCount() == 0) {
+                                    // already found an unsuccessful shard in this index, skip this shard
+                                    return current;
+                                }
+
+                                final ShardSnapshotStatus shardSnapshotStatus = shardEntry.value;
+                                if (shardSnapshotStatus.state() != ShardState.SUCCESS) {
```

Comment:
More of a note to self/observation ... this is a strange state to get to. I wonder if we should just fail the clone overall if any shard failed to clone. We already check that the source shards were successful at the beginning of the clone so if we get here it's due to some unexpected breakage in the repo or some IO issue, either way seems pointless to finalize a broken clone and just write a bunch of extra nonsense to the repo ...

Reply:
This is on the path for snapshotting as well as cloning? Maybe I'm missing something here, I figured this is where we would deal with a partial snapshot of this index.

Reply:
Ah right ... nevermind, we have the issue with snapshots still. The fact that we needlessly get here with broken clone operations remains though but is very unrelated to this PR :) Sorry for the noise.